File: leetcode/data_struct/string/0020/1.go
// auther: phoenix
// date: 2023-06-11 22:26:16
package main

import (
	"fmt"
)

/*


 */

func main() {
	isValid1Test()

}

func isValid1Test() {
	fmt.Println(isValid1("[]"))
	fmt.Println(isValid1("[)"))
	fmt.Println(isValid1("[()()]"))
	fmt.Println(isValid1("[()[]]"))
	fmt.Println(isValid1("[{()[]}]"))
}

func isValid1(s string) bool {
	length := len(s)
	if length == 0 || length%2 == 1 {
		return false
	}

	stack := []byte{}
	pairs := map[byte]byte{
		')': '(',
		'}': '{',
		']': '[',
	}

	for i := 0; i < length; i++ {
		if pairs[s[i]] > 0 {
			if len(stack) == 0 || stack[len(stack)-1] != pairs[s[i]] {
				return false
			}
			stack = stack[:len(stack)-1]
		} else {
			stack = append(stack, s[i])
		}
	}

	return len(stack) == 0
}

File: leetcode/data_struct/string/0020/1.ts
// auther: phoenix
// date: 2023-06-11 22:26:34

/*


*/

demo()
function demo() {
  isValid1Test()
}

function isValid1Test(){
  console.log(isValid1('[]'))
  console.log(isValid1('[)'))
  console.log(isValid1('[()]'))
  console.log(isValid1('[()()]'))
  console.log(isValid1('[()[]]'))
  console.log(isValid1('[{()[]}]'))
}
function isValid1(s:string):boolean{
  const length:number = s.length
  if(length === 0 || length % 2 === 1) {
    return false
  }

  const pairs = new Map([
    [')', '('],
    [']', '['],
    ['}', '{'],
  ])
  const stack:string[] = []
  for (const c of s) {
    if (pairs.has(c)) { // 右侧符号
      if (!stack.length || stack[stack.length - 1] !== pairs.get(c)) { // 栈内无符号 || 栈顶符号与右侧匹配符号不同
        return false
      }
      stack.pop()   // 右侧符号出栈
    } else {

      stack.push(c) // 左侧符号入栈
    }
  }


  return !stack.length
}